lighthouse-core/report/html/renderer/code-snippet-renderer.js

```diff
@@ -0,0 +1,260 @@
+/**
+ * @license Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/* eslint-env jest */
+
+/* globals self, Util */
+
+/** @typedef {import('./details-renderer')} DetailsRenderer */
+
+const SHOW_IF_EXPANDED_CLASS = 'lh-code-snippet__show-if-expanded';
+const SHOW_IF_COLLAPSED_CLASS = 'lh-code-snippet__show-if-collapsed';
+
+/**
+ * @param {LH.Audit.DetailsRendererCodeSnippetLine[]} lines
+ * @param {number} lineNumber
+ */
+function getLineAndPreviousLine(lines, lineNumber) {
+  return {
+    line: lines.find(l => l.number === lineNumber),
+    previousLine: lines.find(l => l.number === lineNumber - 1),
+  };
+}
+
+class CodeSnippetRenderer {
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererCodeSnippet} details
+   * @param {DetailsRenderer} detailsRenderer
+   * @param {function} toggleExpandedFn
+   * @return {DocumentFragment}
+   */
+  static renderHeader(dom, tmpl, details, detailsRenderer, toggleExpandedFn) {
+    const {lineCount, title} = details;
+    const showAll = lineCount <= 4;
+
+    const header = dom.cloneTemplate('#tmpl-lh-code-snippet__header', tmpl);
+    dom.find('.lh-code-snippet__title', header).textContent = title;
+
+    const {codeSnippetCollapse, codeSnippetExpand} = Util.UIStrings;
+    dom.find('.' + SHOW_IF_EXPANDED_CLASS, header).textContent = codeSnippetCollapse;
+    dom.find('.' + SHOW_IF_COLLAPSED_CLASS, header).textContent = codeSnippetExpand;
+
+    const toggleShowAllButton = dom.find('.lh-code-snippet__toggle-show-all', header);
+    if (showAll) {
+      toggleShowAllButton.remove();
+    } else {
+      toggleShowAllButton.addEventListener('click', () => toggleExpandedFn());
```

Comment:
What's the js restriction for the report in devtools again? (We should really write these down :)

I believe we currently don't run any after report generation that's not in devtools itself. Is that still a requirement? If so, we'll either need a decent way to display this statically or land this interaction code in devtools as well.


@paulirish

Reply:
> What's the js restriction for the report in devtools again?

There is none, these days. :)  (But there used to be.)

And yes, as of today, only interaction JS is defined in the DevTools files, but there's nothing preventing us from registering event handlers in our renderers. Seems cool.